### ISSUE TITLE
fido_dev_open is blocked in macOS when called right after fido_dev_info_manifest.

### ISSUE THREAD

**Issue by iradization** [Yubico/libfido2#119]
Hi, 

I'm running the following code while having a Yubico device connected to one of my USB ports. 

```
	fido_dev_info_t *x = fido_dev_info_new(8);
        size_t num = 0;
        fido_dev_info_manifest(x, 8, &num);
        if (num) {
             const fido_dev_info_t *device_info = fido_dev_info_ptr(x, 0);
             const char *device_path = fido_dev_info_path(device_info);
             std::string device_name = fido_dev_info_product_string(device_info);
             fido_dev_t *credential = fido_dev_new();
             fido_dev_open(credential, device_path)); 
              ....
```

However, my process get blocked on `fido_dev_open` as seen from the following sample : 

```
    +           7419 fido_dev_open  (in libfido2.1.3.0.dylib) + 48  [0x10f540c50]
    +             7419 fido_dev_open_wait  (in libfido2.1.3.0.dylib) + 70  [0x10f540cc6]
    +               7419 fido_dev_open_rx  (in libfido2.1.3.0.dylib) + 67  [0x10f5415d3]
    +                 7419 fido_rx  (in libfido2.1.3.0.dylib) + 141  [0x10f545abd]
    +                   7419 rx_preamble  (in libfido2.1.3.0.dylib) + 49  [0x10f545e91]
    +                     7419 rx_frame  (in libfido2.1.3.0.dylib) + 102  [0x10f545f86]
    +                       7419 fido_hid_read  (in libfido2.1.3.0.dylib) + 222  [0x10f54b0be]
    +                         7418 CFRunLoopRunSpecific  (in CoreFoundation) + 455  [0x7fff30ba661e]
    +                         ! 7415 __CFRunLoopRun  (in CoreFoundation) + 1612  [0x7fff30ba6ebc]
    +                         ! : 7415 __CFRunLoopServiceMachPort  (in CoreFoundation) + 328  [0x7fff30ba794e]
    +                         ! :   7415 mach_msg  (in libsystem_kernel.dylib) + 60  [0x7fff5cc3176c]
    +                         ! :     7415 mach_msg_trap  (in libsystem_kernel.dylib) + 10  [0x7fff5cc3122a]
```

In order to resolve the blocking, I need to delay the `fido_dev_open` after calling `fido_dev_info_manifest`. 

Perhaps did you encountered similar behaviour ? 

thanks !

**Comment by martelletto** [comment]
Hi,

This sounds similar to https://github.com/Yubico/libfido2/issues/56#issuecomment-532164571.

I will double-check hid_osx.c's implementation of fido_hid_manifest() to see if we are doing something wrong.

-p.

**Comment by iradization** [comment]
Is there an option to overrun only the method `fido_hid_read`. I'm thinking about setting some timeout to the following loop that fails the method and retry later on : 

```
	do
		r = CFRunLoopRunInMode(dev->loop_id, 0.003, true);
	while (r != kCFRunLoopRunHandledSource);
```

From the examples I only seen that one can overrun the entire methods like this, but I only need one method and would like the rest to remain the same... 

```
typedef struct fido_dev_io {
	fido_dev_io_open_t  *open;
	fido_dev_io_close_t *close;
	fido_dev_io_read_t  *read;
	fido_dev_io_write_t *write;
	fido_dev_io_rx_t    *rx;
	fido_dev_io_tx_t    *tx;
} fido_dev_io_t;

di->io = (fido_dev_io_t) {
		&fido_hid_open,
		&fido_hid_close,
		&fido_hid_read,
		&fido_hid_write,
		NULL,
		NULL,
};
```

**Comment by iradization** [comment]
btw, I'm not sure where does the ~3 seconds delay should be in the code... 

As mentioned, in my case, what solved the problem, is setting a delay between fido_dev_info_new and fido_dev_open... but it was done in my own POC outside libfido API and is definitely not part of macos's USB subsystem as specified in the linked issue. 

I'd like to avoid such constant delays and wait for system event instead which makes the code more robust and readable, but I don't what event to wait for ? 

Perhaps you've found what event should occur to ensure no blocking in fido_hid_read ?

**Comment by martelletto** [comment]
Hi,

Apologies in advance - nothing I am about to say is likely to objectively help you. :(

> Is there an option to overrun only the method `fido_hid_read`. I'm thinking about setting some timeout to the following loop that fails the method and retry later on :
> 
> ```
> 	do
> 		r = CFRunLoopRunInMode(dev->loop_id, 0.003, true);
> 	while (r != kCFRunLoopRunHandledSource);
> ```
> 
> From the examples I only seen that one can overrun the entire methods like this, but I only need one method and would like the rest to remain the same...

My suggestion would be to either link against libfido2 statically or add fido_hid_* to src/export.llvm. You could then cherry-pick a subset of the default HID methods and pass them back to fido_dev_set_io_functions().

A second approach would be to build libfido2 with USE_HIDAPI=1 and see if the delay exists when using hidapi's HID implementation on MacOS.

If you are on Catalina, a third approach would be to explore DriverKit.

Sorry for being unable to help you further.

-p.

**Comment by martelletto** [comment]
Circling back on this. Which Yubico device are you using?

**Comment by martelletto** [comment]
I've pushed a tentative fix to https://github.com/Yubico/libfido2/tree/hid_osx_timeout. I've modified examples/manifest on that branch to continuously loop and poll for keys.

**Comment by martelletto** [comment]
Potentially fixed in 8f8a99d1. Please reopen otherwise. Thank you!
